core/filterconstraint.cpp

```diff
@@ -0,0 +1,939 @@
+// SPDX-License-Identifier: GPL-2.0
+#include "filterconstraint.h"
+#include "dive.h"
+#include "divesite.h"
+#include "errorhelper.h"
+#include "gettextfromc.h"
+#include "qthelper.h"
+#include "tag.h"
+#include "trip.h"
+#include "subsurface-time.h"
+#include <QDateTime>
+
+// We use the units enum only internally.
+// Therefore define it here, not in the header file.
+enum filter_constraint_units {
+	FILTER_CONSTRAINT_NO_UNIT = 0,
+	FILTER_CONSTRAINT_LENGTH_UNIT,
+	FILTER_CONSTRAINT_TEMPERATURE_UNIT,
+	FILTER_CONSTRAINT_WEIGHT_UNIT,
+	FILTER_CONSTRAINT_VOLUMETRIC_FLOW_UNIT
+};
+
+static struct type_description {
+	filter_constraint_type type;
+	const char *s;			// untranslated string that, which will be written to the log
+	const char *translated;
+	bool has_string_mode;
+	bool has_range_mode;
+	bool is_star_widget;
+	filter_constraint_units units;
+	int decimal_places;
+	bool has_date;
+	bool has_time;
+} type_descriptions[] = {
+	{ FILTER_CONSTRAINT_DATE, "date", QT_TRANSLATE_NOOP("gettextFromC", "date"), false, true, false, FILTER_CONSTRAINT_NO_UNIT, 0, true, false },
+	{ FILTER_CONSTRAINT_DATE_TIME, "date_time", QT_TRANSLATE_NOOP("gettextFromC", "date time"), false, true, false, FILTER_CONSTRAINT_NO_UNIT, 0, true, true },
+	{ FILTER_CONSTRAINT_TIME_OF_DAY, "time_of_day", QT_TRANSLATE_NOOP("gettextFromC", "time of day"), false, true, false, FILTER_CONSTRAINT_NO_UNIT, 0, false, true },
+	{ FILTER_CONSTRAINT_YEAR, "year", QT_TRANSLATE_NOOP("gettextFromC", "year"), false, true, false, FILTER_CONSTRAINT_NO_UNIT, 0, false, false },
+	{ FILTER_CONSTRAINT_DAY_OF_WEEK, "day_of_week", QT_TRANSLATE_NOOP("gettextFromC", "week day"), false, false, false, FILTER_CONSTRAINT_NO_UNIT, 0, false, false },
+
```

Comment:
time of day, week, and year, but not month... I'm curious to read on how this will be used...

Reply:
I was contemplating about a "month" constraint. And while "day-of-week" is already stretching it (I implemented it mostly as a proof-of-concept), I could see myself searching for all dives on a Monday, but not so much for all dives in a November. In any case, it should be trivial to add.